ERCS/erc-draft-wallet-connection-api.md

```diff
@@ -0,0 +1,180 @@
+
+## Specification
+
+### `wallet_connect`
+
+Request the user to connect a single account and optionally confirm chain support and add capabilities.
+
+#### RPC Specification
+
+For each chain requested by the app, wallets MUST return a mapped capabilities object if the chain is supported and SHOULD return an empty object if no capabilities exist. Wallets MUST NOT return mapped capabilities objects for chains they do not support. If an app does not declare chains they would like to confirm support for, the wallet can return any chains it wishes to declare support for.
+
+```typescript
+type WalletConnectParams = [{
+  version: string;
+  chains?: `0x${string}`[]; // optional chain IDs (EIP-155 hex)
+  capabilities?: Record<string,any>; // optional connection capabilities
+}]
+
+type WalletConnectResult = {
+  account: {
+    address: `0x${string}`; // connected account address
+    supportedChainsAndCapabilities: Record<`0x${string}`,any>; // chain-specific capabilities, mirrors ERC-5792 wallet_getCapabilities
+  },
+  capabilityResults: Record<string,any>; // results of this connection request's connection capabilities
+}
+```
+
+#### Example Parameters
+
+```json
+[{
+  "version": "1",
+  "chains": ["0x1", "0x2105"],
+  "capabilities": {
+    "exampleCapability": {
+      "foo": "bar"
+    }
+  }
+}]
+```
+
+#### Example Result
+
+```json
+{
+  "account": {
+    "address": "0x...",
+    "supportedChainsAndCapabilities": {
+      "0x1": {},
+      "0x2105": {
+        "atomicBatch": {
+          "supported": true
+        },
+        "paymasterService": {
+          "supported": true
+        }
+      }
+    }
+  },
+  "capabilityResults": {
+    "exampleCapability": {}
+  }
+}
+```
+
+### `signInWithEthereum` Capability
+
+Adds authentication using the [ERC-4361](https://eips.ethereum.org/EIPS/eip-4361) Sign In WIth Ethereum standard.
+
+#### Capability Specification
+
+Same as ERC-4361 specification with minor modifications: 
+* The chain id is optional and if not provided, the wallet SHOULD use the earliest supported chain available from the requested array. 
+* The casing of multi-word fields has been adjusted to mixedCase instead of hyphen-case. Resources are an array field. 
+* The account address returned by `wallet_connect` MUST be the same address that is auto-inserted into the SIWE message.
+
+The wallet MUST return a properly formatted ERC-4361 message that exactly matches the requested parameters and a signature over the EIP-191 hash of the message. The app SHOULD also verify that the two match for security.
+
+```typescript
+type SignInWithEthereumCapabilityParams = {
+  scheme?: string,
+  domain: string,
+  statement?: string
+  uri: string,
+  version: string,
+  chainId?: string
+  nonce: string,
+  issuedAt: string
+  expirationTime?: string,
+  notBefore?: string,
+  requestId?: string,
+  resources?: string[]
+}
+
+type SignInWithEthereumCapabilityResult = {
+  message: string, // formatted SIWE message
```

Comment:
```suggestion
```

Do we need the message? I thought the wallet would be the one displaying the formatted message.

Reply:
Was thinking that the app needs to verify that the message the wallet prompted signing for matches their requested arguments. For instance a malicious wallet could attempt to reuse an old expiry/nonce if it had access to a previously signed SIWE message. The most succinct way to send back all fields and verify against the signature is to pass back the final `message`

Reply:
Got it. Makes sense!